### PR TITLE
Use local PRNG in `spsa_grad`

### PR DESCRIPTION
**Pull request by frederikwilde**
**Context:**
Currently, one can provide an integer as a seed via the `sampler_seed` keyword argument to `qml.gradients.spsa_grad`. Internally, this is passed on to the `sampler` function, which in turn can use it to set the seed for a PRNG. There are three problems with this behaviour and the way it is implemented:

1. The seed in the internal `_rademacher_sampler` is set _globally_ via `np.random.seed()`. This is not advisable, since calls to other functions might also access this global PRNG, which might compromize reproducibility. Even worse, other functions might also set the global seed at any time, which could lead to random samples not having the desired distribution or not being random at all. Setting the global seed is also discouraged by [NumPy](https://numpy.org/doc/stable/reference/random/generated/numpy.random.seed.html)
2. When one wants to use multiple random directions to estimate the gradient, one cannot do this currently, by setting the `sampler_seed` kwarg, since this leads to the same directions being drawn multiple times.
3. When one wants to call `spsa_grad` multiple times (e.g. during gradient descent optimization) in a reproducible way without having the same directions being drawn in each call, one would have to specify a different integer seed in _each_ call.

**Description of the Change:**
The `sampler_seed` kwarg is deprecated and instead the `sampler_rng` kwarg is introduced. This argument can have three different types:

1. `None`: The default, which internally creates a NumPy PRNG without setting a seed, i.e., it is not reproducible.
2. `int`: Is used as a seed for a NumPy PRNG, but in such a way, that when `num_directions` is greater than one, each direction is drawn independently by the `_rademacher_sampler`.
3. `np.random.Generator`: In order to allow users to call `spsa_grad` multiple times without having to specify different seeds each time, a PRNG can be passed explicitly, which is then provided to the sampler.

**Benefits:**
Sampling multiple directions is possible while setting a seed or even a PRNG. By creating one PRNG and passing it to `spsa_grad` a user can easily call `spsa_grad` many times, while keeping the outcomes perfectly reproducible.

**Possible Drawbacks:**
The `sampler_seed` argument is deprecated. When using multiprocessing the NumPy PRNG might not be reproducible. Ideally, one should offer users multiple ways of managing their random states during the computation.

**Related GitHub Issues:**
https://github.com/PennyLaneAI/pennylane/issues/4107

### CODE REVIEW

**Comment by frederikwilde** [comment]
[SC-38157]

File: pennylane/gradients/spsa_gradient.py
Comment (albi3ro) [inline pennylane/gradients/spsa_gradient.py:317]:
```suggestion
    sampler_rng = np.random.default_rng(sampler_rng)
```

See documentation of `default_rng`:

> If passed a [Generator](https://numpy.org/doc/stable/reference/random/generator.html#numpy.random.Generator), it will be returned unaltered.

The initialization logic here can just be handled by `default_rng`.

File: pennylane/gradients/spsa_gradient.py
Comment (albi3ro) [inline pennylane/gradients/spsa_gradient.py:75]: Do we need to rename this keyword argument, as that would be considered a breaking change. Alternatively, we could just expand the definition of  `sampler_seed` to include generators or generally anything `default_rng` accepts.

The big change is in the call signature for the `sampler`.  That's what actually needs to be accepting a generator instead of a seed.

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:317]: Oh, nice!

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:75]: I was also thinking about sticking with the name `sampler_seed`, but I feel like we should encourage users to pass a generator, since they will most likely call `spsa_grad` many times as part of some optimization process. Encouraging them to pass a generator instead of a seed, will probably prevent some users from making the mistake of passing the same seed in each `spsa_grad` call.

But I am also OK with keeping `sampler_seed` and emphasizing the advantage of generators in the documentation.

File: tests/interfaces/test_autograd_qnode_new.py
Comment (frederikwilde) [inline tests/interfaces/test_autograd_qnode_new.py:203]: @albi3ro Do these changes look good to you? There are a lot more of them, but not all of them are exactly the same.

File: pennylane/gradients/spsa_gradient.py
Comment (albi3ro) [inline pennylane/gradients/spsa_gradient.py:584]:
```suggestion
    sampler_rng = np.random.default_rnmg(sampler_rng)
```

Same thing here.

File: tests/interfaces/test_autograd_qnode_new.py
Comment (albi3ro) [reply tests/interfaces/test_autograd_qnode_new.py:203]: Changes are looking good.

Setting a seed in the tests is good for avoiding intermittent failures and reproducible results, but I don't think we need to strictly set it in all of the tests. If we don't set the seed, `spsa` just creates its own random rng, which is perfectly fine.

**Review by albi3ro — COMMENTED** [review]
Let me just double check that I understand what is going on.

It used to be that when we set the seed, the error would stay high no matter how many directions we used, because we just kept using the same directions each time.

Now, we are testing with a specified seed, and we are actually able to get accurate results?

This slipped through the cracks because tests only ever specified a global seed, and never actually used the `sampler_seed` keyword argument?

Yep... Looks like `test_spsa_gradient.py` doesn't actually have any instances of using the `sampler_seed` keyword argument...

**Comment by frederikwilde** [comment]
I did some local tests on the convergence with growing `num_directions`. But actually we should include that in the tests. I'll add some tests for this.

Currently, there is indeed no testing of setting `sampler_seed`. Only in `tests/interfaces/test_jax_jit_qnode.py`, but there `num_directions` is always `1`.

File: tests/interfaces/test_torch_qnode.py
Comment (albi3ro) [inline tests/interfaces/test_torch_qnode.py:499]:
```suggestion
            kwargs = {**kwargs, **spsa_kwargs}
```
Maybe? 

combining dictionaries with `|` is 3.9+, and we still have to support Python 3.8.

File: doc/releases/changelog-dev.md
Comment (frederikwilde) [reply doc/releases/changelog-dev.md]: As discussed with @trbromley a while ago I changed it now to `sampler_seed` being deprecated.

File: doc/development/deprecations.rst
Comment (dwierichs) [inline doc/development/deprecations.rst:138]:
```suggestion
  to create a PRNG internally, or to a NumPy pseudo-random number generator (PRNG) created via
```

File: doc/development/deprecations.rst
Comment (dwierichs) [inline doc/development/deprecations.rst:141]:
```suggestion
  multiple times, for instance during an optimization procedure.
```

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [inline pennylane/gradients/spsa_gradient.py:307]:
```suggestion
        raise ValueError(f"The argument sampler_rng is expected to be a NumPy PRNG, an integer or None, but is {sampler_rng}.")
```

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [inline pennylane/gradients/spsa_gradient.py:325]: Just an idea:

```suggestion
    if sampler_seed is not None:
        msg = (
            "The sampler_seed argument is deprecated. Use the sampler_rng argument "
            "instead to pass a random number generator or a seed."
        )
        if sampler_rng is None:
            warnings.warn(msg, UserWarning)
            sampler_rng = sampler_seed
        else:
            raise ValueError("Both sampler_rng and sampler_seed were specified. Only specify one.\n" + msg)
```

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [inline pennylane/gradients/spsa_gradient.py:609]: Makes me think whether we want a `process_sampler_args` method to be called in both versions of `_spsa_grad` :)

File: tests/gradients/finite_diff/test_spsa_gradient.py
Comment (dwierichs) [inline tests/gradients/finite_diff/test_spsa_gradient.py:64]:
```suggestion
        err = "Both sampler_rng and sampler_seed were specified."
```

File: tests/gradients/finite_diff/test_spsa_gradient.py
Comment (dwierichs) [inline tests/gradients/finite_diff/test_spsa_gradient.py:45]: Needs a docstring :)

**Review by dwierichs — COMMENTED** [review]
Thanks @frederikwilde, this makes the seed control so much better! :) :100:

I just had a few minor remarks, and a question about the `rademacher_sampler` unit tests.
I agree with @albi3ro that renaming the keyword argument might be nice to avoid. However, I also see how extending `sampler_seed` to take PRNGs may be a bit unintuitive...

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:39]: Since this function is not exposed externally and is guaranteed to receive a `np.random.Generator` in `spsa_grad()` I'd suggest making this a required (positional) argument. If we want to make it a kwarg, we'd have to make sure that it cannot be an integer, since this was the problem which initially lead to this PR, where the same integer was passed each time `_rademacher_sampler` was called.

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:307]: Should we do `"... but is {type(sampler_rng)}."` instead?

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:609]: I think we are getting read of the legacy return type system in the next release cycle, no? Otherwise this would definitely make sense!

File: tests/gradients/finite_diff/test_spsa_gradient.py
Comment (frederikwilde) [inline tests/gradients/finite_diff/test_spsa_gradient.py:44]:
```suggestion
    def test_sampler_seed_deprecation(self):
        """
        Ensure that passing the sampler_seed kwarg results in a deprecation warning
        and cannot be combined with the new sampler_rng kwarg.
        """
```

**Review by albi3ro — APPROVED** [review]
Looks like there's now a test failing due to error messages being slightly different, but I'm happy to approve anyway as long as that gets fixed before the PR is merged.

Glad you were able to track down the source of the failing tests :)

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:307]: I was thinking so too. I'm not sure what's the more helpful situation here. One reason I saw against printing the `type` was that an object of type `A` can be an instance of a parent class `B`, detaching the printed info a bit from the actual test that leads to the error :thinking: 
However, it probably does not matter too much and if you find printing the `type` more clear, go for it :)

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:602]: Yep, true. Still could be nice to factor out. Not crucial though, feel free to just ignore :)

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:39]: Ah, right. It's true, the function is private. Maybe we just want to make `rng` a required argument _always_, including in `coordinate_sampler` and in the descriptions in the docs? This way we avoid users copying and modifying the sampler and running into this trap? :thinking:

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:307]: Oh, that's a good point. I'll add it as you suggested.

File: tests/legacy/test_legacy_spsa_gradient_old.py
Comment (frederikwilde) [reply tests/legacy/test_legacy_spsa_gradient_old.py:49]: I copied the `TestRademacherSampler` class into the new tests. They all pass. Not sure if there is any changes necessary to un-legacify these tests.

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:39]: I agree that this argument should be non-optional, since it's a sampler. But where would you change it in the docs? I only see the `coordinate_sampler` being defined in tests.

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [inline pennylane/gradients/spsa_gradient.py:129]: Mention of `rng` being a keyword argument

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [inline pennylane/gradients/spsa_gradient.py:457]: Mention of `rng` being a keyword argument

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py:39]: Ah, sorry, could have been more clear on that. I marked two points where `rng` is explicitly called a keyword argument :)

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py]: Technically, in `spsa_grad` the `rng` is passed as a keyword argument. And in the `_rademacher_sampler` it is also defined as a _required_ keyword argument. So the `sampler` can accept `rng` as anything that is not a "positional only" argument. I made this more precise in the docs and also added a test for this. https://github.com/PennyLaneAI/pennylane/pull/4165/commits/2eda609429086619842e12da08ccf3a14c68a647

File: pennylane/gradients/spsa_gradient.py
Comment (frederikwilde) [reply pennylane/gradients/spsa_gradient.py:39]: I see. Thanks!

File: pennylane/gradients/spsa_gradient.py
Comment (dwierichs) [reply pennylane/gradients/spsa_gradient.py]: Great, thanks for the thorough clarification :)

**Review by dwierichs — APPROVED** [review]
Thanks a lot @frederikwilde :)

**Pull request merged**
